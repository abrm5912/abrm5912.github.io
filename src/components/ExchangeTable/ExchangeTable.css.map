{"version":3,"sources":["ExchangeTable.less"],"names":[],"mappings":"AAAA;EACE,mBAAA;;AADF,eAEE;EACE,gBAAA;EACA,gBAAA;EACA,cAAA;;AAIJ,QAA0B,uBAAuB;EAC/C;IACE,uBAAA;;;AAIJ,QAA0B,uBAAuB;EAC/C;IACE,uBAAA;;;AAIJ,QAA2B,wBAAwB;EACjD;IACE,uBAAA;;;AAIJ,QAA2B;EACzB;IACE,oBAAA","file":"ExchangeTable.css"}