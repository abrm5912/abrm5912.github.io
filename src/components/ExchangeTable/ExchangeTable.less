.exchange-table {
  margin: 0 auto 50px;
  .ant-table {
    font-size: 1vmax;
    max-height: 60vh;
    overflow: auto;
  }
}

@media (min-width: 360px) and (max-width: 415px) {
  .exchange-table {
    margin: 220px 50px 50px;
  }
}

@media (min-width: 415px) and (max-width: 800px) {
  .exchange-table {
    margin: 100px 50px 50px;
  }
}

@media (min-height: 360px) and (max-height: 800px) {
  .exchange-table {
    margin: 400px 50px 50px;
  }
}

@media (min-width: 1500px) {
  .exchange-table {
    margin: 0 500px 50px;
  }
}