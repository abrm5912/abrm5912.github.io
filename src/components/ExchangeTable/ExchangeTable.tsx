import React, { FC, useEffect, useState } from 'react';
import { Table } from 'antd';
import { EditableRow } from './EditableRow/EditableRow';
import { EditableCell } from './EditableCell/EditableCell';
import { useExchangeStore } from '../../store';
import { ColumnTypes } from '../../types/table';
import { ExchangeDataFromAPI } from '../../types/exchange';
import { useEditableTable } from '../../hooks/useEditableTable';
import { defaultColumns } from '../../constants/table';
import './ExchangeTable.css'

interface ExchangeDataType {
  key: number;
  currency: string;
  buy: number;
  sale: number;
}

interface ExchangeTableProps {
  data?: ExchangeDataFromAPI[];
}

export const ExchangeTable: FC<ExchangeTableProps> = ({ data }) => {
  const exchangeItems = useExchangeStore(state => state.items);
  const setExchangeItems = useExchangeStore(state => state.setExchangeItems);
  const updateExchangeItem = useExchangeStore(state => state.updateExchangeItem);
  const isLoading = useExchangeStore(state => state.isLoading);
  const [dataSource, setDataSource] = useState<ExchangeDataType[]>([]);

  const { columns, editedItem } = useEditableTable<ExchangeDataType>(defaultColumns, dataSource, setDataSource);

  useEffect(() => {
    if (data) {
      setExchangeItems(data.map((item: ExchangeDataFromAPI, index: number) => ({ ...item, id: index + 1 })));
    }
  }, [data]);

  useEffect(() => {
    if (!dataSource.length) {
      setDataSource(exchangeItems.map((item) => ({
        key: item.id,
        currency: `${item.ccy}/${item.base_ccy}`,
        buy: item.buy,
        sale: item.sale
      })));
    }
  }, [exchangeItems]);

  useEffect(() => {
    if (editedItem) {
      const exchangeItem = exchangeItems.find((item) => item.id === editedItem.key);
      if (exchangeItem) {
        updateExchangeItem({
          ...exchangeItem,
          buy: editedItem.buy,
          sale: editedItem.sale
        });
      }
    }
  }, [editedItem]);

  const components = {
    body: {
      row: EditableRow,
      cell: EditableCell,
    },
  };

  return (
    <Table
      components={components}
      bordered
      tableLayout="auto"
      rowClassName={() => 'editable-row'}
      loading={isLoading}
      className="exchange-table"
      data-testid="exchange-table"
      pagination={false}
      columns={columns as ColumnTypes}
      dataSource={dataSource}
    />
  );
};
