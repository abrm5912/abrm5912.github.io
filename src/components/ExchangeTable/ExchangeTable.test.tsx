import { render } from '@testing-library/react';
import React from 'react';
import { ExchangeTable } from './ExchangeTable';

global.matchMedia = global.matchMedia || function () {
  return {
    addListener: jest.fn(),
    removeListener: jest.fn(),
  };
};

const mockData = [
  { ccy: 'USD', base_ccy: 'UAH', buy: 27.80000, sale: 28.20000 },
  { ccy: 'EUR', base_ccy: 'UAH', buy: 32.50000, sale: 33.20000 },
  { ccy: 'RUR', base_ccy: 'UAH', buy: 0.30000, sale: 0.39000 },
  { ccy: 'BTC', base_ccy: 'USD', buy: 10000, sale: 11000 },
];

describe('ExchangeTable', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.resetModules();
  });

  it('should render without error', () => {
    const { container } = render(<ExchangeTable data={mockData}/>);

    expect(container).toMatchSnapshot();
  });

  it('should render without data', () => {
    const { container } = render(<ExchangeTable data={undefined}/>);

    expect(container).toMatchSnapshot();
  });

  it('should render with empty data', () => {
    const { container } = render(<ExchangeTable data={[]}/>);

    expect(container).toMatchSnapshot();
  });
});
