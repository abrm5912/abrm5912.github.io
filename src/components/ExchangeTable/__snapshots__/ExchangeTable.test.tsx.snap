// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangeTable should render with empty data 1`] = `
<div>
  <div
    class="ant-table-wrapper exchange-table css-dev-only-do-not-override-nllxry"
  >
    <div
      class="ant-spin-nested-loading css-dev-only-do-not-override-nllxry"
    >
      <div
        class="ant-spin-container"
      >
        <div
          class="ant-table ant-table-bordered ant-table-empty"
          data-testid="exchange-table"
        >
          <div
            class="ant-table-container"
          >
            <div
              class="ant-table-content"
            >
              <table
                style="table-layout: auto;"
              >
                <colgroup>
                  <col
                    style="width: 10%;"
                  />
                  <col
                    style="width: 150px;"
                  />
                  <col
                    style="width: 150px;"
                  />
                </colgroup>
                <thead
                  class="ant-table-thead"
                >
                  <tr>
                    <th
                      class="ant-table-cell"
                      scope="col"
                      style="text-align: center;"
                    >
                      Currency
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                      style="text-align: center;"
                    >
                      Buy
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                      style="text-align: center;"
                    >
                      Sell
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="ant-table-tbody"
                >
                  <tr
                    class="ant-table-placeholder"
                  >
                    <td
                      class="ant-table-cell"
                      colspan="3"
                    >
                      <div
                        class="css-dev-only-do-not-override-nllxry ant-empty ant-empty-normal"
                      >
                        <div
                          class="ant-empty-image"
                        >
                          <svg
                            height="41"
                            viewBox="0 0 64 41"
                            width="64"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <g
                              fill="none"
                              fill-rule="evenodd"
                              transform="translate(0 1)"
                            >
                              <ellipse
                                cx="32"
                                cy="33"
                                fill="#f5f5f5"
                                rx="32"
                                ry="7"
                              />
                              <g
                                fill-rule="nonzero"
                                stroke="#d9d9d9"
                              >
                                <path
                                  d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                />
                                <path
                                  d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                  fill="#fafafa"
                                />
                              </g>
                            </g>
                          </svg>
                        </div>
                        <div
                          class="ant-empty-description"
                        >
                          No data
                        </div>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ExchangeTable should render without data 1`] = `
<div>
  <div
    class="ant-table-wrapper exchange-table css-dev-only-do-not-override-nllxry"
  >
    <div
      class="ant-spin-nested-loading css-dev-only-do-not-override-nllxry"
    >
      <div
        class="ant-spin-container"
      >
        <div
          class="ant-table ant-table-bordered"
          data-testid="exchange-table"
        >
          <div
            class="ant-table-container"
          >
            <div
              class="ant-table-content"
            >
              <table
                style="table-layout: auto;"
              >
                <colgroup>
                  <col
                    style="width: 10%;"
                  />
                  <col
                    style="width: 150px;"
                  />
                  <col
                    style="width: 150px;"
                  />
                </colgroup>
                <thead
                  class="ant-table-thead"
                >
                  <tr>
                    <th
                      class="ant-table-cell"
                      scope="col"
                      style="text-align: center;"
                    >
                      Currency
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                      style="text-align: center;"
                    >
                      Buy
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                      style="text-align: center;"
                    >
                      Sell
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="ant-table-tbody"
                >
                  <tr
                    class="ant-table-row ant-table-row-level-0 editable-row"
                    data-row-key="1"
                  >
                    <td
                      class="ant-table-cell"
                      style="text-align: center;"
                    >
                      USD/UAH
                    </td>
                    <td
                      class="ant-table-cell"
                      style="text-align: center;"
                    >
                      <div
                        class="editable-cell-value-wrap"
                      >
                        27.8
                        <svg
                          class="edit-icon"
                        >
                          edit.svg
                        </svg>
                      </div>
                    </td>
                    <td
                      class="ant-table-cell"
                      style="text-align: center;"
                    >
                      <div
                        class="editable-cell-value-wrap"
                      >
                        28.2
                        <svg
                          class="edit-icon"
                        >
                          edit.svg
                        </svg>
                      </div>
                    </td>
                  </tr>
                  <tr
                    class="ant-table-row ant-table-row-level-0 editable-row"
                    data-row-key="2"
                  >
                    <td
                      class="ant-table-cell"
                      style="text-align: center;"
                    >
                      EUR/UAH
                    </td>
                    <td
                      class="ant-table-cell"
                      style="text-align: center;"
                    >
                      <div
                        class="editable-cell-value-wrap"
                      >
                        32.5
                        <svg
                          class="edit-icon"
                        >
                          edit.svg
                        </svg>
                      </div>
                    </td>
                    <td
                      class="ant-table-cell"
                      style="text-align: center;"
                    >
                      <div
                        class="editable-cell-value-wrap"
                      >
                        33.2
                        <svg
                          class="edit-icon"
                        >
                          edit.svg
                        </svg>
                      </div>
                    </td>
                  </tr>
                  <tr
                    class="ant-table-row ant-table-row-level-0 editable-row"
                    data-row-key="3"
                  >
                    <td
                      class="ant-table-cell"
                      style="text-align: center;"
                    >
                      RUR/UAH
                    </td>
                    <td
                      class="ant-table-cell"
                      style="text-align: center;"
                    >
                      <div
                        class="editable-cell-value-wrap"
                      >
                        0.3
                        <svg
                          class="edit-icon"
                        >
                          edit.svg
                        </svg>
                      </div>
                    </td>
                    <td
                      class="ant-table-cell"
                      style="text-align: center;"
                    >
                      <div
                        class="editable-cell-value-wrap"
                      >
                        0.39
                        <svg
                          class="edit-icon"
                        >
                          edit.svg
                        </svg>
                      </div>
                    </td>
                  </tr>
                  <tr
                    class="ant-table-row ant-table-row-level-0 editable-row"
                    data-row-key="4"
                  >
                    <td
                      class="ant-table-cell"
                      style="text-align: center;"
                    >
                      BTC/USD
                    </td>
                    <td
                      class="ant-table-cell"
                      style="text-align: center;"
                    >
                      <div
                        class="editable-cell-value-wrap"
                      >
                        10000
                        <svg
                          class="edit-icon"
                        >
                          edit.svg
                        </svg>
                      </div>
                    </td>
                    <td
                      class="ant-table-cell"
                      style="text-align: center;"
                    >
                      <div
                        class="editable-cell-value-wrap"
                      >
                        11000
                        <svg
                          class="edit-icon"
                        >
                          edit.svg
                        </svg>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ExchangeTable should render without error 1`] = `
<div>
  <div
    class="ant-table-wrapper exchange-table css-dev-only-do-not-override-nllxry"
  >
    <div
      class="ant-spin-nested-loading css-dev-only-do-not-override-nllxry"
    >
      <div
        class="ant-spin-container"
      >
        <div
          class="ant-table ant-table-bordered"
          data-testid="exchange-table"
        >
          <div
            class="ant-table-container"
          >
            <div
              class="ant-table-content"
            >
              <table
                style="table-layout: auto;"
              >
                <colgroup>
                  <col
                    style="width: 10%;"
                  />
                  <col
                    style="width: 150px;"
                  />
                  <col
                    style="width: 150px;"
                  />
                </colgroup>
                <thead
                  class="ant-table-thead"
                >
                  <tr>
                    <th
                      class="ant-table-cell"
                      scope="col"
                      style="text-align: center;"
                    >
                      Currency
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                      style="text-align: center;"
                    >
                      Buy
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                      style="text-align: center;"
                    >
                      Sell
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="ant-table-tbody"
                >
                  <tr
                    class="ant-table-row ant-table-row-level-0 editable-row"
                    data-row-key="1"
                  >
                    <td
                      class="ant-table-cell"
                      style="text-align: center;"
                    >
                      USD/UAH
                    </td>
                    <td
                      class="ant-table-cell"
                      style="text-align: center;"
                    >
                      <div
                        class="editable-cell-value-wrap"
                      >
                        27.8
                        <svg
                          class="edit-icon"
                        >
                          edit.svg
                        </svg>
                      </div>
                    </td>
                    <td
                      class="ant-table-cell"
                      style="text-align: center;"
                    >
                      <div
                        class="editable-cell-value-wrap"
                      >
                        28.2
                        <svg
                          class="edit-icon"
                        >
                          edit.svg
                        </svg>
                      </div>
                    </td>
                  </tr>
                  <tr
                    class="ant-table-row ant-table-row-level-0 editable-row"
                    data-row-key="2"
                  >
                    <td
                      class="ant-table-cell"
                      style="text-align: center;"
                    >
                      EUR/UAH
                    </td>
                    <td
                      class="ant-table-cell"
                      style="text-align: center;"
                    >
                      <div
                        class="editable-cell-value-wrap"
                      >
                        32.5
                        <svg
                          class="edit-icon"
                        >
                          edit.svg
                        </svg>
                      </div>
                    </td>
                    <td
                      class="ant-table-cell"
                      style="text-align: center;"
                    >
                      <div
                        class="editable-cell-value-wrap"
                      >
                        33.2
                        <svg
                          class="edit-icon"
                        >
                          edit.svg
                        </svg>
                      </div>
                    </td>
                  </tr>
                  <tr
                    class="ant-table-row ant-table-row-level-0 editable-row"
                    data-row-key="3"
                  >
                    <td
                      class="ant-table-cell"
                      style="text-align: center;"
                    >
                      RUR/UAH
                    </td>
                    <td
                      class="ant-table-cell"
                      style="text-align: center;"
                    >
                      <div
                        class="editable-cell-value-wrap"
                      >
                        0.3
                        <svg
                          class="edit-icon"
                        >
                          edit.svg
                        </svg>
                      </div>
                    </td>
                    <td
                      class="ant-table-cell"
                      style="text-align: center;"
                    >
                      <div
                        class="editable-cell-value-wrap"
                      >
                        0.39
                        <svg
                          class="edit-icon"
                        >
                          edit.svg
                        </svg>
                      </div>
                    </td>
                  </tr>
                  <tr
                    class="ant-table-row ant-table-row-level-0 editable-row"
                    data-row-key="4"
                  >
                    <td
                      class="ant-table-cell"
                      style="text-align: center;"
                    >
                      BTC/USD
                    </td>
                    <td
                      class="ant-table-cell"
                      style="text-align: center;"
                    >
                      <div
                        class="editable-cell-value-wrap"
                      >
                        10000
                        <svg
                          class="edit-icon"
                        >
                          edit.svg
                        </svg>
                      </div>
                    </td>
                    <td
                      class="ant-table-cell"
                      style="text-align: center;"
                    >
                      <div
                        class="editable-cell-value-wrap"
                      >
                        11000
                        <svg
                          class="edit-icon"
                        >
                          edit.svg
                        </svg>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
